First pass standalone NFS (#176)
hpe-storage/scod: docs/csi_driver/standalone_nfs.md

# Standalone NFS Server

In certain situations is desirable to run the NFS Server Provisioner image without the dual `PersistentVolumeClaim` (PVC) semantic in a more static fashion on top of a `PVC` provisioned by a non-HPE CSI Driver `StorageClass`.

[TOC]

## Limitations

> - The standalone NFS server is not part of the HPE CSI Driver and should be considered a standalone Kubernetes application altogether. The HPE CSI Driver NFS Server Provisioner NFS servers may co-exist on the same cluster and `Namespace` without risk of conflict but not recommended.

Reviewer: ```suggestion
- The standalone NFS server is not part of the HPE CSI Driver and is deployed as a standalone Kubernetes application. The HPE CSI Driver NFS Server Provisioner and standalone NFS server instances may co-exist on the same cluster and `Namespace` without risk of conflict but is not recommended.
```

> - The [Pod Monitor](monitor.md) which normally monitors `Pods` status for the "NodeLost" condition is not included with the standalone NFS server and recovery is at the mercy of the underlying storage platform and driver.

Reviewer: ```suggestion
- The [Pod Monitor](monitor.md) which normally monitors `Pods` status for "NodeLost" conditions is not included with the standalone NFS server and recovery is the responsibility of the underlying storage platform and driver.
```

> - Support is limited on the standalone NFS server and only available to select users.

Reviewer: Support by whom? HPE?

Who are the "select users"? Certain HPE customers?

## Prerequisites

> It's assumed during the creation steps that a Kubernetes cluster is available with enough permissions to deploy privileged containers with `SYS_ADMIN` and `DAC_READ_SEARCH` capabilities. All steps are run in a terminal with `kubectl` and `git` in in the path.

Reviewer: ```suggestion
It's assumed the user deploying the standalone NFS server has appropriate permissions to deploy privileged containers with `SYS_ADMIN` and `DAC_READ_SEARCH` capabilities on the respective Kubernetes cluster. All steps are run in a terminal with `kubectl` and `git` in in the path.
```

- A default `StorageClass` declared on the cluster
- Worker nodes that will serve the NFS exports *must* be labeled with `csi.hpe.com/hpe-nfs: "true"`
- `kubectl` and Kubernetes v1.21 or newer

## Create a Workspace

NFS server configurations are managed with the kustomize templating system. Clone this repository to get started and change working directory.

```text
git clone https://github.com/hpe-storage/scod
cd scod/docs/csi_driver/examples/standalone_nfs
```

In the current directory, various manifests and configuration directives exist to deploy and manage NFS servers.

Run `tree .` in the current directory: 

```text
.
├── base
│   ├── configmap.yaml
│   ├── deployment.yaml
│   ├── environment.properties
│   ├── kustomization.yaml
│   ├── pvc.yaml
│   ├── service.yaml
│   └── values.yaml
└── overlays
    └── example
        ├── environment.properties
        └── kustomization.yaml

4 directories, 9 files
```

!!! important 
    The current directory is now the "home" for the remainder of this guide.

## Create an NFS Server

> Copy the "example" overlay into a new directory. In the examples "my-server" is used.

Reviewer: ```suggestion
Copy the "example" overlay into a new directory. For this guide, the folder "my-server" is used.
```

```text
cp -a overlays/example overlays/my-server
```

Edit both "environment.properties" and "kustomization.yaml"

### environment.properties

```text
# This is the domain associated with worker node (not inter-cluster DNS)
CLUSTER_NODE_DOMAIN_NAME=my-domain.example.com

# The size of the backend RWO claim
PERSISTENCE_SIZE=16Gi

# Default resource limits for the NFS server
NFS_SERVER_CPU_LIMIT=1
NFS_SERVER_MEMORY_LIMIT=2Gi
```

The "CLUSTER_NODE_DOMAIN_NAME" variable refers to the DNS domain name that the worker node is resolvable in, not the Kubernetes cluster DNS.

> The "PERSISTENCE_SIZE" is the backend `PVC` size expressed in the same format accepted by a `PVC`.

Reviewer: "the same format", do we have the acceptible SI units documented somewhere for reference? (M, G, T)

service/my-server-hpe-nfs created
persistentvolumeclaim/my-server-hpe-nfs created
deployment.apps/my-server-hpe-nfs created
```

Inspect the resources with `kubectl get -k overlays/my-server`:

```text
NAME                                        DATA   AGE
configmap/my-server-hpe-nfs-conf            1      59s
configmap/my-server-local-conf-97898bftbh   2      59s

NAME                        TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                                                                                                               AGE
service/my-server-hpe-nfs   ClusterIP   10.100.200.11   <none>        49000/TCP,2049/TCP,2049/UDP,32803/TCP,32803/UDP,20048/TCP,20048/UDP,111/TCP,111/UDP,662/TCP,662/UDP,875/TCP,875/UDP   59s

NAME                                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
persistentvolumeclaim/my-server-hpe-nfs   Bound    pvc-ae943116-d0af-4696-8b1b-1dcf4316bdc2   18Gi       RWO            vsphere-sc     58s

NAME                                READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/my-server-hpe-nfs   1/1     1            1           59s
```

> Make a note of the IP address assigned to "service/my-server-hpe-nfs", that is the IP address needed to mount the NFS export.

Reviewer: ```suggestion
Make a note of the IP address assigned to "service/my-server-hpe-nfs", that is the IP address that will be used to mount NFS export(s).
```

!!! tip
    If the Kubernetes cluster DNS service is resolvable from the worker node host OS, it possible to use the cluster DNS address to mount the `Service`, in this example that would be "my-server-hpe-nfs.default.svc.cluster.local".

## Mounting the NFS Server

There are two ways to mount the NFS server.

> 1. Inline declaration of where to find the NFS server and NFS Export

Reviewer: ```suggestion
1. Inline declaration of the NFS server and NFS Export
```
Is "Export" a proper name?

1. Statically creating a `PersistentVolume` with the NFS server details and mount options and manually claiming the `PV` with a `PVC` using the `.spec.volumeName` parameter

### Inline Declaration

> This is the most elegant solution as it does not require any intermediary `PVC` or `PV` and directly refers to the NFS server within a workload stanza.

Reviewer: ```suggestion
This is the most elegant solution as it does not require any intermediary `PVC` or `PV` and defines the NFS server within a workload stanza.
```

> This is an example from a `StatefulSet` workload controller having multiple replicas.

Reviewer: ```suggestion
This is an example of a `StatefulSet` workload controller with multiple replicas.
```

```text
...
spec:
  replicas: 3
  template:
    ...
    spec:
      containers:
        volumeMounts:
        - name: vol
          mountPath: /vol
      ...
      volumes:
      - name: vol
        nfs:
          server: 10.100.200.11
          path: /export
```

!!! important
    Replace `.spec.template.spec.volumes[].nfs.server` with IP address from the actual `Service` IP address and not the examples.

### Static Provisioning

Refer to the [official Kubernetes documentation](https://kubernetes.io/docs/concepts/storage/volumes/#nfs) for the built-in NFS client on how to perform static provisioning of NFS `PVs` and `PVCs`.

## Expand PVC

If the `StorageClass` and underlying CSI driver supports volume expansion, simply edit `overlays/my-server/environment.properties` with the new (larger) size and issue `kubectl apply -k overlays/my-server` to expand the volume.

## Deleting the NFS Server

> Ensure no workloads have active mounts against the NFS server `Service`. If there are, those `Pods` will be stuck indefinitely. 

Reviewer: ```suggestion
Ensure there are no workloads with active mounts to the NFS server `Service`. If the mounts are deleted while `Pods` are still active, those `Pods` will be stuck indefinitely. 
```
